File: go/oop-patterns/structural-patterns/decorator/decorator.go
package main

import "fmt"

//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
// Decorator (ДЕКОРАТОР)
// ЦЕЛЬ: расширить функционал конкретного объекта, не меняя при этом логики
// класса (схемы), из которого был создан объект

// Интерфейс объекта, который оборачивается
// ему должны так же следовать все обёртки, чтобы клиентский код
// мог без труда использовать столько декораторов, сколько угодно
type iDecorated interface {
	log(data string)
}

// Конкретный объект со стандартной имплементацией операций
type decorated struct {
}

func (d *decorated) log(data string) {
	fmt.Println(data)
}

// Базовый декоратор должен так же следовать интерфейсу оборачиваемого объекта
type decorator struct {
	decorated iDecorated
}

func (d *decorator) log(data string) {
	d.decorated.log(data)
}

// Конкретные декораторы следуют интерфейсу оборачиваемого объекта и
// манипулируют данными, передаваемыми в него или получаемыми от него
type concreteDecoratorA struct {
	decorator
}

func (d *concreteDecoratorA) log(data string) {
	data = fmt.Sprintf("Декоратор А преобразовал данные, строка до преобразования:\n%v", data)
	d.decorator.log(data)
}

func newConcreteDecoratorA(d iDecorated) *concreteDecoratorA {
	return &concreteDecoratorA{decorator{decorated: d}}
}

type concreteDecoratorB struct {
	decorator
}

func (d *concreteDecoratorB) log(data string) {
	data = fmt.Sprintf("Декоратор B преобразовал данные, строка до преобразования:\n%v", data)
	d.decorator.log(data)
}

func newConcreteDecoratorB(d iDecorated) *concreteDecoratorB {
	return &concreteDecoratorB{decorator{decorated: d}}
}

// decorated := &decorated{}
// decoratorA := newConcreteDecoratorA(decorated)
// decoratorB := newConcreteDecoratorB(decoratorA)

// decorated.log("Привет!")
// decoratorA.log("Привет!")
// decoratorB.log("Привет!")

File: typescript/oop-patterns/structural-patterns/decorator/decorator.ts
// Интерфейс объекта, который оборачивается
// ему должны так же следовать все обёртки, чтобы клиентский код
// мог без труда использовать столько декораторов, сколько угодно
interface IDecorated {
  log(data: string): void;
}

// Конкретный класс со стандартной имплементацией операций
class Decorated implements IDecorated {
  public log(data: string): void {
    console.log(data);
  }
}

// Базовый декоратор должен так же следовать интерфейсу оборачиваемого объекта
class Decorator implements IDecorated {
  protected decorated: IDecorated;

  constructor(decorated: IDecorated) {
    this.decorated = decorated;
  }
  public log(data: string): void {
    this.decorated.log(data);
  }
}

// Конкретные декораторы следуют интерфейсу оборачиваемого объекта и
// манипулируют данными, передаваемыми в него или получаемыми от него
class ConcreteDecoratorA extends Decorator {
  public log(data: string): void {
    data = `Декоратор А преобразовал данные, строка до преобразования:\n${data}`;
    super.log(data);
  }
}

class ConcreteDecoratorB extends Decorator {
  public log(data: string): void {
    data = `Декоратор B преобразовал данные, строка до преобразования:\n${data}`;
    super.log(data);
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

const decorated = new Decorated();
const decoratorA = new ConcreteDecoratorA(decorated);
const decoratorB = new ConcreteDecoratorB(decoratorA);

decorated.log('Привет!');
decoratorA.log('Привет!');
decoratorB.log('Привет!');
